nixos/modules/services/networking/iwd.nix

```diff
@@ -4,16 +4,142 @@ with lib;
+  };
+
+  # The file extension of the network-config determines the type of the network. Currently,
+  # we support public networks (`open`) and networks protected with a passphrase (`psk`).
+  # Anything else, like 802.1x networks are not supported by this module.
+  determineType = cfg: if cfg.psk != null || cfg.passphrase != null then "psk" else "open";
+
+  # Generates a name-value-pair with configurations for a SSID. For further information
+  # about the configuration options, please refer to the upstream docs:
+  #
+  # https://iwd.wiki.kernel.org/networkconfigurationsettings
+  predefinedNetworks = flip mapAttrs cfg.networks (name: cfg: {
+    extension = determineType cfg;
+
+    # Cannot use the SSID here: the hex-encode is done at build-time, not eval time
+    # and several special chars in a file name would break `writeText`.
+    file = pkgs.writeText "ssid" (generators.toINI {} ({
+      Settings.AutoConnect = cfg.autoConnect;
+      Settings.Hidden = cfg.hidden;
+    } // (optionalAttrs (cfg.psk != null) {
+      Security.PreSharedKey = cfg.psk;
+    }) // (optionalAttrs (cfg.passphrase != null) {
+      Security.Passprase = cfg.passphrase;
+    })));
+  });
 in {
-  options.networking.wireless.iwd.enable = mkEnableOption "iwd";
+  options.networking.wireless.iwd = {
+    enable = mkEnableOption "iwd";
+
+    interfaces = mkOption {
+      type = types.nullOr (types.listOf types.str);
+      example = [ "wlp2s0" ];
+      apply = x: if x == null then x else concatStringsSep " " x;
+      default = null;
+      description = ''
+        List of interfaces to use.
+      '';
+    };
+
+    networks = mkOption {
+      default = {};
+      example = {
+        "karlsruhe.freifunk.net" = {};
+        "secured".passphrase = "12345678";
+        "network-with-psk".psk = "bf4c086e055e492373a0f59ff614f61fc2d597a56db94e17a165f9ef1d42c066";
+      };
+
+      description = ''
+        Declarative configuration of wifi networks for
+        <citerefentry><refentrytitle>iwd</refentrytitle><manvolnum>8</manvolnum></citerefentry>.
+
+        All networks will be stored in
+        <literal>/var/lib/iwd/&lt;name&gt;.&lt;type&gt;</literal>.
+
+        Since each network is stored in its own file, declarative networks can be used in an
+        environment with imperatively added networks via
+        <citerefentry><refentrytitle>iwctl</refentrytitle><manvolnum>1</manvolnum></citerefentry>.
+      '';
+
+      type = types.attrsOf (types.submodule {
+        options = {
+          psk = mkOption {
+            type = types.nullOr (fixedStringLength 64 64);
+            default = null;
+            description = ''
+              WPA PSK for the currently defined network. This can be generated by using
+              <citerefentry><refentrytitle>wpa_passphrase</refentrytitle><manvolnum>8</manvolnum></citerefentry>.
+            '';
+          };
+
+          passphrase = mkOption {
```

Comment:
And manually constructing the INI file during runtime has the benefit that you can convert the `psk` and `passphrase` options to `pskFile` and `passphraseFile` respectively, which then won't force the user into putting their password in the Nix store. When constructing the INI file in `preStart` you can read these file paths to get the secret values.

Note that if people don't care about this, they can still do `psk = pkgs.writeFile "foo-psk" "the-psk"`, which still puts the secret in the store.

Furthermore, 802.1x support can be added similarly, as requested by @onny.

Reply:
Relevant issue: https://github.com/NixOS/nixpkgs/issues/24288